lib/src/optimism/mod.rs

```diff
@@ -0,0 +1,348 @@
+                    .context("failed to create batcher transactions")?;
+
+                self.eth_block_no += 1;
+            }
+
+            // Process batches
+            while let Some(op_batch) = self.op_batcher.read_batch()? {
+                // Process the batch
+                self.op_block_no += 1;
+
+                #[cfg(not(target_os = "zkvm"))]
+                info!(
+                    "Read batch for Op block {}: timestamp={}, epoch={}, tx count={}, parent hash={:?}",
+                    self.op_block_no,
+                    op_batch.essence.timestamp,
+                    op_batch.essence.epoch_num,
+                    op_batch.essence.transactions.len(),
+                    op_batch.essence.parent_hash,
+                );
+
+                // Update sequence number (and fetch deposits if start of new epoch)
+                let deposits =
+                    if op_batch.essence.epoch_num == self.op_batcher.state.epoch.number + 1 {
+                        self.op_block_seq_no = 0;
+                        self.op_batcher.state.do_next_epoch()?;
+
+                        self.op_batcher
+                            .state
+                            .epoch
+                            .deposits
+                            .iter()
+                            .map(|tx| tx.to_rlp())
+                            .collect()
+                    } else {
+                        self.op_block_seq_no += 1;
+
+                        Vec::new()
+                    };
+
+                // Obtain new Op head
+                let new_op_head = {
+                    let new_op_head = self
+                        .derive_input
+                        .db
+                        .get_op_block_header(self.op_block_no)
+                        .context("block not found")?;
+
+                    // Verify new op head has the expected parent
+                    assert_eq!(
+                        new_op_head.parent_hash,
+                        self.op_batcher.state.safe_head.hash
+                    );
+
+                    // Verify that the new op head transactions are consistent with the batch
+                    // transactions
+                    {
+                        let system_tx = self.derive_system_transaction(&op_batch);
+
+                        let derived_transactions: Vec<_> = once(system_tx.to_rlp())
+                            .chain(deposits)
+                            .chain(op_batch.essence.transactions.iter().map(|tx| tx.to_vec()))
+                            .collect();
+
+                        let mut tx_trie = MptNode::default();
+                        for (tx_no, tx) in derived_transactions.into_iter().enumerate() {
+                            let trie_key = tx_no.to_rlp();
+                            tx_trie.insert(&trie_key, tx)?;
+                        }
+                        if tx_trie.hash() != new_op_head.transactions_root {
```

Comment:
Can't we validate more fields of the new_op_head ?

Reply:
My thinking is that Zeth verifies the other fields. (Specifically: if Zeth agrees that the block hash of `new_op_head` is valid, then all of the fields are valid.) This, together with a `derive` proof that the transaction list is valid, completes the proof.